golfscript.rb

```diff
@@ -2,153 +2,116 @@
+	def uniop(rhs); rhs.is_a?(Numeric) ? self | rhs : (a,b=gscoerce(rhs); a.uniop(b)); end
+	def intop(rhs); rhs.is_a?(Numeric) ? self & rhs : (a,b=gscoerce(rhs); a.intop(b)); end
+	def difop(rhs); rhs.is_a?(Numeric) ? self ^ rhs : (a,b=gscoerce(rhs); a.difop(b)); end
+	def to_gs; Gstring.new(to_s); end
+	def ginspect; to_gs; end
+	def go; $stack<<self; end
+	def notop; self == 0 ? 1 : 0; end
 	def class_id; 0; end
-	def coerce(b)
-		[if b.class == Garray
+	def falsey; self == 0; end
+	def ltop(rhs); self < rhs ? 1 : 0; end
+	def gtop(rhs); self > rhs ? 1 : 0; end
+	def equalop(rhs); self == rhs ? 1 : 0; end
+	def question(b); self**b; end
+	def leftparen; self-1; end
+	def rightparen; self+1; end
+	def gscoerce(b)
+		c = b.class_id
+		[if c == 1
 			Garray.new([self])
-		elsif b.class == Gstring
+		elsif c == 2
 			to_gs
 		else #Gblock
 			to_gs.to_s.compile
 		end,b]
 	end
-	
-	def ~
-		Gint.new(~@val)
-	end
-	def notop
-		Gint.new(@val == 0)
-	end
-	'*/%<>'.each_byte{|i|
-		eval'def %c(rhs)
-			Gint.new(@val %c rhs.val)
-		end'%[i,i]
-	}
-	def equalop(rhs)
-		Gint.new(@val == rhs.val)
-	end
-	def question(b)
-		Gint.new(@val**b.val)
-	end
 	def base(a)
 		if Garray===a
 			r=0
 			a.val.each{|i|
-				r*=@val
-				r+=i.val
+				r*=self
+				r+=i
 			}
-			Gint.new(r)
+			r
 		else
-			i=a.val.abs
+			i=a.abs
 			r=[]
 			while i!=0
-				r.unshift Gint.new(i%@val)
-				i/=@val
+				r.unshift(i % self)
+				i/=self
 			end
 			Garray.new(r)
 		end
 	end
-	def leftparen
-		Gint.new(@val-1)
-	end
-	def rightparen
-		Gint.new(@val+1)
-	end
 end
 
 class Garray < Gtype
 	def initialize(a)
 		@val = a || []
 	end
+	def concat(rhs)
+		if rhs.class != self.class
+			a,b=gscoerce(rhs)
+			a.addop(b)
+		else
+			@val.concat(rhs.val)
+			self
+		end
+	end
 	def factory(a)
 		Garray.new(a)
 	end
 	def to_gs
-		@val.inject(Gstring.new("")){|s,i|s+i.to_gs}
-	end
-	def flatten #maybe name to_a ?		
-# 		Garray.new(@val.inject([]){|s,i|s+case i
-# 			when Gstring then i.val
-# 			when Gint then [i]
-# 			when Garray then i.flatten.val
-# 			when Gblock then i.val
-# 			end
-# 		})
-# 	end
-		#use Peter Taylor's fix to avoid quadratic flatten times
+		r = []
+		@val.each {|i| r.concat(i.to_gs.val) }
+		Gstring.new(r)
+	end
+	def flatten
 		Garray.new(flatten_append([]))
 	end
 	def flatten_append(prefix)
 		@val.inject(prefix){|s,i|case i
-			when Gint then s<<i
+			when Numeric then s<<i
 			when Garray then i.flatten_append(s)
 			when Gstring then s.concat(i.val)
 			when Gblock then s.concat(i.val)
  			end
 		}
 	end
 	def ginspect
-		Gstring.new('[')+Garray.new(@val.map{|i|i.ginspect})*Gstring.new(' ')+Gstring.new(']')
-	end
-	def go
-		$stack<<self
+		bs = []
+		@val.each{|i| bs << 32; bs.concat(i.ginspect.val) }
+		bs[0] = 91
+		bs << 93
+		Gstring.new(bs)
 	end
+	def go; $stack<<self; end
 	def class_id; 1; end
-	def coerce(b)
-		if b.class == Gint
-			b.coerce(self).reverse
-		elsif b.class == Gstring
+	def gscoerce(b)
```

Comment:
I had to rename `coerce` to `gscoerce` because there is already a method named `coerce` on Ruby's `Numeric` and it was messing things up.